syntax = "proto3";

package org.couchers.admin;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

import "annotations.proto";
import "api.proto";
import "communities.proto";

service Admin {
  option (auth_level) = AUTH_LEVEL_ADMIN;

  rpc GetUserDetails(GetUserDetailsReq) returns (UserDetails) {}

  rpc GetUser(GetUserReq) returns (org.couchers.api.core.User) {
    // Get info about a particular user, ignores the fact that they may be banned/deleted
  }

  rpc ChangeUserGender(ChangeUserGenderReq) returns (UserDetails) {}

  rpc ChangeUserBirthdate(ChangeUserBirthdateReq) returns (UserDetails) {}

  rpc AddBadge(AddBadgeReq) returns (UserDetails) {}

  rpc RemoveBadge(RemoveBadgeReq) returns (UserDetails) {}

  rpc SetPassportSexGenderException(SetPassportSexGenderExceptionReq) returns (UserDetails) {}

  rpc BanUser(BanUserReq) returns (UserDetails) {}

  rpc UnbanUser(UnbanUserReq) returns (UserDetails) {}

  rpc AddAdminNote(AddAdminNoteReq) returns (UserDetails) {}

  rpc GetContentReport(GetContentReportReq) returns (GetContentReportRes) {}

  rpc GetContentReportsForAuthor(GetContentReportsForAuthorReq) returns (GetContentReportsForAuthorRes) {}

  rpc SendModNote(SendModNoteReq) returns (UserDetails) {}

  rpc DeleteUser(DeleteUserReq) returns (UserDetails) {}

  rpc RecoverDeletedUser(RecoverDeletedUserReq) returns (UserDetails) {}

  rpc CreateApiKey(CreateApiKeyReq) returns (CreateApiKeyRes) {
    // Create an API key for a user. For security, we don't return the API key to the admin calling this API, rather
    // it's sent to the user in an email
  }

  rpc CreateCommunity(CreateCommunityReq) returns (org.couchers.api.communities.Community) {}

  rpc UpdateCommunity(UpdateCommunityReq) returns (org.couchers.api.communities.Community) {}

  rpc GetChats(GetChatsReq) returns (GetChatsRes) {}

  rpc ListEventCommunityInviteRequests(ListEventCommunityInviteRequestsReq) returns (ListEventCommunityInviteRequestsRes) {}

  rpc DecideEventCommunityInviteRequest(DecideEventCommunityInviteRequestReq) returns (DecideEventCommunityInviteRequestRes) {}

  rpc DeleteEvent(DeleteEventReq) returns (google.protobuf.Empty) {}

  rpc ListUserIds(ListUserIdsReq) returns (ListUserIdsRes) {}

  rpc EditReferenceText(EditReferenceTextReq) returns (google.protobuf.Empty) {}

  rpc DeleteReference(DeleteReferenceReq) returns (google.protobuf.Empty) {}

  rpc EditDiscussion(EditDiscussionReq) returns (google.protobuf.Empty) {}

  rpc EditReply(EditReplyReq) returns (google.protobuf.Empty) {}

  rpc CreateAccountDeletionLink(CreateAccountDeletionLinkReq) returns (CreateAccountDeletionLinkRes) {}

  rpc AccessStats(AccessStatsReq) returns (AccessStatsRes) {}
}

message UserDetails {
  int64 user_id = 1;
  string username = 2;
  string name = 14;
  string email = 3;
  string gender = 4;
  string birthdate = 7;
  bool banned = 5;
  bool deleted = 6;
  bool do_not_email = 15;
  repeated string badges = 8;

  bool has_strong_verification = 9;
  org.couchers.api.core.BirthdateVerificationStatus birthdate_verification_status = 10;
  org.couchers.api.core.GenderVerificationStatus gender_verification_status = 11;
  bool has_passport_sex_gender_exception = 12;

  string admin_note = 13;

  int64 pending_mod_notes_count = 16;
  int64 acknowledged_mod_notes_count =17;
}

message GetUserDetailsReq {
  // username, email, or user id
  string user = 1;
}

message GetUserReq {
  string user = 1;
}

message ChangeUserGenderReq {
  // username, email, or user id
  string user = 1;
  string gender = 2;
}

message ChangeUserBirthdateReq {
  // username, email, or user id
  string user = 1;
  string birthdate = 2; // in YYYY-MM-DD format
}

message AddBadgeReq {
  string user = 1;
  string badge_id = 2;
}

message RemoveBadgeReq {
  string user = 1;
  string badge_id = 2;
}

message SetPassportSexGenderExceptionReq {
  string user = 1;
  bool passport_sex_gender_exception = 2;
}

message BanUserReq {
  // username, email, or user id
  string user = 1;
  string admin_note = 2;
}

message UnbanUserReq {
  // username, email, or user id
  string user = 1;
  string admin_note = 2;
}

message AddAdminNoteReq {
  // username, email, or user id
  string user = 1;
  string admin_note = 2;
}

message ContentReport {
  int64 content_report_id = 1;
  google.protobuf.Timestamp time = 2;
  int64 reporting_user_id = 3;
  int64 author_user_id = 4;
  string reason = 5;
  string description = 6;
  string content_ref = 7;
  string user_agent = 8;
  string page = 9;
}

message GetContentReportReq {
  int64 content_report_id = 1;
}

message GetContentReportRes {
  ContentReport content_report = 1;
}

message GetContentReportsForAuthorReq {
  string user = 1;
}

message GetContentReportsForAuthorRes {
  repeated ContentReport content_reports = 1;
}

message SendModNoteReq {
  string user = 1;
  string content = 2; // CommonMark without images
  // this allows us to track what type of note it is without having to read it individually
  string internal_id = 3;
  // whether to not notify the user about the note: simply show it to them the next time they log on
  bool do_not_notify = 4;
}

message DeleteUserReq {
  // username, email, or user id
  string user = 1;
}

message RecoverDeletedUserReq {
  // username, email, or user id
  string user = 1;
}

message CreateApiKeyReq {
  // username, email, or user id
  string user = 1;
}

message CreateApiKeyRes {}

message CreateCommunityReq {
  string name = 1;
  string description = 3;
  int64 parent_node_id = 4;
  repeated int64 admin_ids = 5;
  string geojson = 6;
}

message UpdateCommunityReq {
  int64 community_id = 1;
  string name = 2;
  string description = 3;
  int64 parent_node_id = 4;
  string geojson = 5;
}

message GetChatsReq {
  // username, email, or user id
  string user = 1;
}

message GetChatsRes {
  string response = 2;
}

message ListEventCommunityInviteRequestsReq {
  uint32 page_size = 1;
  string page_token = 2;
}

message EventCommunityInviteRequest {
  int64 event_community_invite_request_id = 1;

  int64 user_id = 2;
  string event_url = 3;

  // assuming they have notifs enabled
  uint32 approx_users_to_notify = 4;
  int64 community_id = 5;
}

message ListEventCommunityInviteRequestsRes {
  repeated EventCommunityInviteRequest requests = 1;

  string next_page_token = 2;
}

message DecideEventCommunityInviteRequestReq {
  int64 event_community_invite_request_id = 1;

  // whether to approve or deny it
  bool approve = 2;
}

message DecideEventCommunityInviteRequestRes {}

message DeleteEventReq {
  int64 event_id = 1;
}

message ListUserIdsReq {
  google.protobuf.Timestamp start_time = 1;
  google.protobuf.Timestamp end_time = 2;

  uint32 page_size = 3;
  string page_token = 4;
}

message ListUserIdsRes {
  repeated int64 user_ids = 1;

  string next_page_token = 2;
}

message EditReferenceTextReq {
  int64 reference_id = 1;
  string new_text = 2;
}

message DeleteReferenceReq {
  int64 reference_id = 1;
}

message EditDiscussionReq {
  int64 discussion_id = 1;
  string new_title = 2;
  string new_content = 3;
}

message EditReplyReq {
  int64 reply_id = 1;
  string new_content = 2;
}

message CreateAccountDeletionLinkReq {
  // username, email, or user id
  string user = 1;
  int64 expiry_days = 2;
}

message CreateAccountDeletionLinkRes {
  string account_deletion_confirm_url = 1;
}

message AccessStatsReq {
  // username, email, or user id
  string user = 1;
  google.protobuf.Timestamp start_time = 2;
  google.protobuf.Timestamp end_time = 3;
}

message AccessStat {
  string ip_address = 1;
  string asn = 2;
  string asorg = 3;
  string asnetwork = 4;
  string user_agent = 5;
  string operating_system = 6;
  string browser = 7;
  string device = 8;
  string approximate_location = 9;
  uint64 api_call_count = 10;
  uint64 periods_count = 11;
  google.protobuf.Timestamp first_seen = 12;
  google.protobuf.Timestamp last_seen = 13;
}

message AccessStatsRes {
  repeated AccessStat stats = 1;
}
